src/data/enum.lean

```diff
@@ -0,0 +1,190 @@
+
+import category.monad.basic
+import data.list.basic
+import data.equiv.basic
+import data.finset
+import data.fintype
+
+open finset (hiding singleton)
+
+namespace fin
+
+def enum (n : ℕ) : list (fin n) :=
+(list.range n).attach.map $ λ ⟨i,h⟩, ⟨i,by simpa using h⟩
+
+@[simp] lemma mem_enum {n} (x : fin n) : x ∈ enum n :=
+by simp [enum]; existsi x.1; simp [fin.is_lt]
+
+end fin
+
+class enumerable (α : Sort*) :=
```

Comment:
I'm not keen on the name. To my ear `nat` is certainly enumerable. Is it too tedious to write `finitely_enumerable`?

Reply:
I agree that's a better name. It is a mouthful though. How about `fin_enum`?

Reply:
Sounds good to me.